import sys
from PyQt5.QtWidgets import QMessageBox, QMainWindow, QWidget, QDialog, QStyledItemDelegate
from PyQt5 import QtCore, QtGui, QtWidgets
from source import advanced_prediction
import machine_learning


class Ui_predict_window(QMainWindow):
    def __init__(self, max_depth, number, parent=None):
        self.max_depth = max_depth
        self.number = number
        super(Ui_predict_window, self).__init__(parent)

    def setupUi(self):
        self.setObjectName("MainWindow")
        self.resize(1310, 730)
        font = QtGui.QFont()
        font.setPointSize(10)
        self.setFont(font)
        self.centralwidget = QtWidgets.QWidget(self)
        self.setWindowIcon(QtGui.QIcon("../icon/logo.png"))
        self.centralwidget.setObjectName("centralwidget")
        self.label = QtWidgets.QLabel(self.centralwidget)
        self.label.setGeometry(QtCore.QRect(640, 10, 101, 31))
        font = QtGui.QFont()
        font.setPointSize(18)
        self.label.setFont(font)
        self.label.setObjectName("label")
        self.label_2 = QtWidgets.QLabel(self.centralwidget)
        self.label_2.setGeometry(QtCore.QRect(20, 70, 111, 31))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.label_2.setFont(font)
        self.label_2.setObjectName("label_2")
        self.label_3 = QtWidgets.QLabel(self.centralwidget)
        self.label_3.setGeometry(QtCore.QRect(20, 120, 141, 31))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.label_3.setFont(font)
        self.label_3.setObjectName("label_3")
        self.d2 = QtWidgets.QLineEdit(self.centralwidget)
        self.d2.setGeometry(QtCore.QRect(190, 120, 113, 22))
        self.d2.setObjectName("d2")
        self.d3 = QtWidgets.QLineEdit(self.centralwidget)
        self.d3.setGeometry(QtCore.QRect(190, 170, 113, 22))
        self.d3.setObjectName("d3")
        self.label_4 = QtWidgets.QLabel(self.centralwidget)
        self.label_4.setGeometry(QtCore.QRect(20, 170, 131, 31))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.label_4.setFont(font)
        self.label_4.setObjectName("label_4")
        self.label_5 = QtWidgets.QLabel(self.centralwidget)
        self.label_5.setGeometry(QtCore.QRect(30, 230, 71, 16))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.label_5.setFont(font)
        self.label_5.setObjectName("label_5")
        self.d4 = QtWidgets.QLineEdit(self.centralwidget)
        self.d4.setGeometry(QtCore.QRect(190, 220, 113, 22))
        self.d4.setObjectName("d4")
        self.label_6 = QtWidgets.QLabel(self.centralwidget)
        self.label_6.setGeometry(QtCore.QRect(20, 270, 101, 21))
        font = QtGui.QFont()
        font.setPointSize(10)
        self.label_6.setFont(font)
        self.label_6.setObjectName("label_6")
        self.c5 = QtWidgets.QComboBox(self.centralwidget)
        self.c5.setGeometry(QtCore.QRect(190, 270, 111, 21))
        self.c5.setObjectName("c5")
        self.c5.addItem("")
        self.c5.addItem("")
        self.label_7 = QtWidgets.QLabel(self.centralwidget)
        self.label_7.setGeometry(QtCore.QRect(20, 320, 161, 31))
        self.label_7.setObjectName("label_7")
        self.d6 = QtWidgets.QLineEdit(self.centralwidget)
        self.d6.setGeometry(QtCore.QRect(190, 320, 113, 22))
        self.d6.setObjectName("d6")
        self.label_9 = QtWidgets.QLabel(self.centralwidget)
        self.label_9.setGeometry(QtCore.QRect(20, 370, 191, 31))
        self.label_9.setObjectName("label_9")
        self.d7 = QtWidgets.QLineEdit(self.centralwidget)
        self.d7.setGeometry(QtCore.QRect(190, 370, 113, 22))
        self.d7.setObjectName("d7")
        self.label_10 = QtWidgets.QLabel(self.centralwidget)
        self.label_10.setGeometry(QtCore.QRect(20, 420, 171, 21))
        self.label_10.setObjectName("label_10")
        self.d8 = QtWidgets.QLineEdit(self.centralwidget)
        self.d8.setGeometry(QtCore.QRect(190, 420, 113, 22))
        self.d8.setObjectName("d8")
        self.label_11 = QtWidgets.QLabel(self.centralwidget)
        self.label_11.setGeometry(QtCore.QRect(20, 470, 161, 21))
        self.label_11.setObjectName("label_11")
        self.d9 = QtWidgets.QLineEdit(self.centralwidget)
        self.d9.setGeometry(QtCore.QRect(190, 470, 113, 22))
        self.d9.setObjectName("d9")
        self.label_12 = QtWidgets.QLabel(self.centralwidget)
        self.label_12.setGeometry(QtCore.QRect(20, 520, 161, 21))
        self.label_12.setObjectName("label_12")
        self.d10 = QtWidgets.QLineEdit(self.centralwidget)
        self.d10.setGeometry(QtCore.QRect(190, 520, 113, 22))
        self.d10.setObjectName("d10")
        self.label_13 = QtWidgets.QLabel(self.centralwidget)
        self.label_13.setGeometry(QtCore.QRect(20, 570, 81, 21))
        self.label_13.setObjectName("label_13")
        self.label_14 = QtWidgets.QLabel(self.centralwidget)
        self.label_14.setGeometry(QtCore.QRect(350, 70, 201, 21))
        self.label_14.setObjectName("label_14")
        self.label_15 = QtWidgets.QLabel(self.centralwidget)
        self.label_15.setGeometry(QtCore.QRect(350, 120, 55, 16))
        self.label_15.setObjectName("label_15")
        self.label_16 = QtWidgets.QLabel(self.centralwidget)
        self.label_16.setGeometry(QtCore.QRect(350, 170, 181, 21))
        self.label_16.setObjectName("label_16")
        self.label_17 = QtWidgets.QLabel(self.centralwidget)
        self.label_17.setGeometry(QtCore.QRect(350, 220, 61, 16))
        self.label_17.setObjectName("label_17")
        self.label_18 = QtWidgets.QLabel(self.centralwidget)
        self.label_18.setGeometry(QtCore.QRect(350, 270, 151, 16))
        self.label_18.setObjectName("label_18")
        self.label_19 = QtWidgets.QLabel(self.centralwidget)
        self.label_19.setGeometry(QtCore.QRect(350, 320, 151, 21))
        self.label_19.setObjectName("label_19")
        self.c17 = QtWidgets.QLineEdit(self.centralwidget)
        self.c17.setGeometry(QtCore.QRect(650, 320, 113, 22))
        self.c17.setObjectName("c17")
        self.label_21 = QtWidgets.QLabel(self.centralwidget)
        self.label_21.setGeometry(QtCore.QRect(350, 370, 271, 16))
        self.label_21.setObjectName("label_21")
        self.c18 = QtWidgets.QLineEdit(self.centralwidget)
        self.c18.setGeometry(QtCore.QRect(650, 370, 113, 22))
        self.c18.setObjectName("c18")
        self.label_22 = QtWidgets.QLabel(self.centralwidget)
        self.label_22.setGeometry(QtCore.QRect(350, 420, 271, 16))
        self.label_22.setObjectName("label_22")
        self.c19 = QtWidgets.QLineEdit(self.centralwidget)
        self.c19.setGeometry(QtCore.QRect(650, 420, 113, 22))
        self.c19.setObjectName("c19")
        self.label_23 = QtWidgets.QLabel(self.centralwidget)
        self.label_23.setGeometry(QtCore.QRect(350, 470, 291, 16))
        self.label_23.setObjectName("label_23")
        self.c20 = QtWidgets.QLineEdit(self.centralwidget)
        self.c20.setGeometry(QtCore.QRect(650, 470, 113, 22))
        self.c20.setObjectName("c20")
        self.label_24 = QtWidgets.QLabel(self.centralwidget)
        self.label_24.setGeometry(QtCore.QRect(350, 520, 271, 21))
        self.label_24.setObjectName("label_24")
        self.c21 = QtWidgets.QLineEdit(self.centralwidget)
        self.c21.setGeometry(QtCore.QRect(650, 520, 113, 22))
        self.c21.setObjectName("c21")
        self.label_25 = QtWidgets.QLabel(self.centralwidget)
        self.label_25.setGeometry(QtCore.QRect(350, 570, 251, 21))
        self.label_25.setObjectName("label_25")
        self.c22 = QtWidgets.QLineEdit(self.centralwidget)
        self.c22.setGeometry(QtCore.QRect(650, 570, 113, 22))
        self.c22.setObjectName("c22")
        self.label_26 = QtWidgets.QLabel(self.centralwidget)
        self.label_26.setGeometry(QtCore.QRect(800, 70, 351, 16))
        self.label_26.setObjectName("label_26")
        self.c23 = QtWidgets.QLineEdit(self.centralwidget)
        self.c23.setGeometry(QtCore.QRect(1160, 60, 113, 22))
        self.c23.setObjectName("c23")
        self.label_27 = QtWidgets.QLabel(self.centralwidget)
        self.label_27.setGeometry(QtCore.QRect(800, 120, 271, 16))
        self.label_27.setObjectName("label_27")
        self.c24 = QtWidgets.QLineEdit(self.centralwidget)
        self.c24.setGeometry(QtCore.QRect(1160, 110, 113, 22))
        self.c24.setObjectName("c24")
        self.label_28 = QtWidgets.QLabel(self.centralwidget)
        self.label_28.setGeometry(QtCore.QRect(800, 170, 271, 16))
        self.label_28.setObjectName("label_28")
        self.c25 = QtWidgets.QLineEdit(self.centralwidget)
        self.c25.setGeometry(QtCore.QRect(1160, 160, 113, 22))
        self.c25.setObjectName("c25")
        self.label_29 = QtWidgets.QLabel(self.centralwidget)
        self.label_29.setGeometry(QtCore.QRect(800, 220, 281, 16))
        self.label_29.setObjectName("label_29")
        self.c26 = QtWidgets.QLineEdit(self.centralwidget)
        self.c26.setGeometry(QtCore.QRect(1160, 210, 113, 22))
        self.c26.setObjectName("c26")
        self.label_30 = QtWidgets.QLabel(self.centralwidget)
        self.label_30.setGeometry(QtCore.QRect(800, 270, 291, 16))
        self.label_30.setObjectName("label_30")
        self.c27 = QtWidgets.QLineEdit(self.centralwidget)
        self.c27.setGeometry(QtCore.QRect(1160, 260, 113, 22))
        self.c27.setObjectName("c27")
        self.label_31 = QtWidgets.QLabel(self.centralwidget)
        self.label_31.setGeometry(QtCore.QRect(800, 320, 281, 16))
        self.label_31.setObjectName("label_31")
        self.c28 = QtWidgets.QLineEdit(self.centralwidget)
        self.c28.setGeometry(QtCore.QRect(1160, 310, 113, 22))
        self.c28.setObjectName("c28")
        self.label_32 = QtWidgets.QLabel(self.centralwidget)
        self.label_32.setGeometry(QtCore.QRect(800, 370, 341, 16))
        self.label_32.setObjectName("label_32")
        self.c29 = QtWidgets.QLineEdit(self.centralwidget)
        self.c29.setGeometry(QtCore.QRect(1160, 360, 113, 22))
        self.c29.setObjectName("c29")
        self.label_33 = QtWidgets.QLabel(self.centralwidget)
        self.label_33.setGeometry(QtCore.QRect(800, 420, 191, 16))
        self.label_33.setObjectName("label_33")
        self.c30 = QtWidgets.QLineEdit(self.centralwidget)
        self.c30.setGeometry(QtCore.QRect(1160, 410, 113, 22))
        self.c30.setObjectName("c30")
        self.label_34 = QtWidgets.QLabel(self.centralwidget)
        self.label_34.setGeometry(QtCore.QRect(800, 520, 47, 13))
        self.label_34.setObjectName("label_34")
        self.c32 = QtWidgets.QLineEdit(self.centralwidget)
        self.c32.setGeometry(QtCore.QRect(1160, 510, 113, 22))
        self.c32.setObjectName("c32")
        # self.label_35 = QtWidgets.QLabel(self.centralwidget)
        # self.label_35.setGeometry(QtCore.QRect(800, 500, 61, 21))
        # self.label_35.setObjectName("label_35")
        self.predict_button = QtWidgets.QPushButton(self.centralwidget)
        self.predict_button.setStyleSheet("background-color:rgb(94, 186, 125)")
        self.predict_button.clicked.connect(self.predict)
        self.predict_button.setIcon(QtGui.QIcon('../icon/predict.png'))
        self.predict_button.setGeometry(QtCore.QRect(460, 650, 93, 28))
        self.predict_button.setObjectName("predict_button")
        self.cancel_btn = QtWidgets.QPushButton(self.centralwidget)
        self.cancel_btn.setStyleSheet("background-color:rgb(218, 91, 11)")
        self.cancel_btn.clicked.connect(self.cancel)
        self.cancel_btn.setIcon(QtGui.QIcon('../icon/cancel.png'))
        self.cancel_btn.setGeometry(QtCore.QRect(840, 650, 93, 28))
        self.cancel_btn.setObjectName("cancel_btn")
        self.label_link = QtWidgets.QLabel(self.centralwidget)
        self.label_link.setGeometry(QtCore.QRect(800, 500, 61, 21))
        self.label_link.hide()
        self.label_link.setObjectName("label_link")
        self.label_tt = QtWidgets.QLabel(self.centralwidget)
        self.label_tt.hide()
        self.label_tt.setGeometry(QtCore.QRect(800, 500, 61, 21))
        self.label_tt.setObjectName("label_tt")

        self.c1 = QtWidgets.QComboBox(self.centralwidget)
        self.c1.setGeometry(QtCore.QRect(190, 70, 111, 21))
        self.c1.setObjectName("c1")
        self.c1.addItem("")
        self.c1.addItem("")
        self.c11 = QtWidgets.QComboBox(self.centralwidget)
        self.c11.setGeometry(QtCore.QRect(190, 570, 111, 22))
        self.c11.setObjectName("c11")
        self.c11.addItem("")
        self.c11.addItem("")
        self.c12 = QtWidgets.QComboBox(self.centralwidget)
        self.c12.setGeometry(QtCore.QRect(650, 70, 111, 22))
        self.c12.setObjectName("c12")
        self.c12.addItem("")
        self.c12.addItem("")
        self.c13 = QtWidgets.QComboBox(self.centralwidget)
        self.c13.setGeometry(QtCore.QRect(650, 120, 111, 22))
        self.c13.setObjectName("c13")
        self.c13.addItem("")
        self.c13.addItem("")
        self.c14 = QtWidgets.QComboBox(self.centralwidget)
        self.c14.setGeometry(QtCore.QRect(650, 170, 111, 22))
        self.c14.setObjectName("c14")
        self.c14.addItem("")
        self.c14.addItem("")
        self.c16 = QtWidgets.QComboBox(self.centralwidget)
        self.c16.setGeometry(QtCore.QRect(650, 270, 111, 22))
        self.c16.setObjectName("c16")
        self.c16.addItem("")
        self.c16.addItem("")
        self.c15 = QtWidgets.QComboBox(self.centralwidget)
        self.c15.setGeometry(QtCore.QRect(650, 220, 111, 22))
        self.c15.setObjectName("c15")
        self.c15.addItem("")
        self.c15.addItem("")
        self.label_35 = QtWidgets.QLabel(self.centralwidget)
        self.label_35.setGeometry(QtCore.QRect(800, 470, 121, 16))
        self.label_35.setObjectName("label_35")
        self.c31 = QtWidgets.QLineEdit(self.centralwidget)
        self.c31.setGeometry(QtCore.QRect(1160, 460, 113, 22))
        self.c31.setObjectName("c31")
        self.predict_button_2 = QtWidgets.QPushButton(self.centralwidget)
        self.predict_button_2.setStyleSheet("background-color:rgb(94, 186, 125)")
        self.predict_button_2.clicked.connect(self.ad_predict)
        self.predict_button_2.setIcon(QtGui.QIcon('../icon/advanced.png'))
        self.predict_button_2.setGeometry(QtCore.QRect(630, 650, 160, 28))
        self.predict_button_2.setObjectName("predict_button_2")
        self.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(self)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 1310, 25))
        self.menubar.setObjectName("menubar")
        self.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(self)
        self.statusbar.setObjectName("statusbar")
        self.setStatusBar(self.statusbar)

        self.retranslateUi()

    def retranslateUi(self):
        _translate = QtCore.QCoreApplication.translate
        self.setWindowTitle(_translate("MainWindow", "MainWindow"))
        self.label.setText(_translate("MainWindow", "Predict"))
        self.label_2.setText(_translate("MainWindow", "Marital status:"))
        self.label_3.setText(_translate("MainWindow", "Application mode: "))
        self.label_4.setText(_translate("MainWindow", "Application order:"))
        self.label_5.setText(_translate("MainWindow", "Course:"))
        self.label_6.setText(_translate("MainWindow", " Attendance:"))
        self.c5.setItemText(0, _translate("MainWindow", "Morning"))
        self.c5.setItemText(1, _translate("MainWindow", "Evening"))
        self.label_7.setText(_translate("MainWindow", "Previous qualification:"))
        self.label_9.setText(_translate("MainWindow", "Mother\'s qualification:"))
        self.label_10.setText(_translate("MainWindow", "Father\'s qualification:"))
        self.label_11.setText(_translate("MainWindow", "Mother\'s occupation:"))
        self.label_12.setText(_translate("MainWindow", "Father\'s occupation:"))
        self.label_13.setText(_translate("MainWindow", "Displaced:"))
        self.label_14.setText(_translate("MainWindow", "Educational special needs:"))
        self.label_15.setText(_translate("MainWindow", "Debtor:"))
        self.label_16.setText(_translate("MainWindow", "Tuition fees up to date:"))
        self.label_17.setText(_translate("MainWindow", "Gender:"))
        self.label_18.setText(_translate("MainWindow", "Scholarship holder:"))
        self.label_19.setText(_translate("MainWindow", "Age at enrollment:"))
        self.label_21.setText(_translate("MainWindow", "Curricular units 1st sem (credited):"))
        self.label_22.setText(_translate("MainWindow", "Curricular units 1st sem (enrolled):"))
        self.label_23.setText(_translate("MainWindow", "Curricular units 1st sem (evaluations):"))
        self.label_24.setText(_translate("MainWindow", "Curricular units 1st sem (approved):"))
        self.label_25.setText(_translate("MainWindow", "Curricular units 1st sem (grade):"))
        self.label_26.setText(_translate("MainWindow", "Curricular units 1st sem (without evaluations):"))
        self.label_27.setText(_translate("MainWindow", "Curricular units 2nd sem (credited):"))
        self.label_28.setText(_translate("MainWindow", "Curricular units 2nd sem (enrolled):"))
        self.label_29.setText(_translate("MainWindow", "Curricular units 2nd sem (evaluations):"))
        self.label_30.setText(_translate("MainWindow", "Curricular units 2nd sem (approved):"))
        self.label_31.setText(_translate("MainWindow", "Curricular units 2nd sem (grade):"))
        self.label_32.setText(_translate("MainWindow", "Curricular units 2nd sem (without evaluations):"))
        self.label_33.setText(_translate("MainWindow", "Unemployment rate:"))
        self.label_34.setText(_translate("MainWindow", "GDP:"))
        self.label_link.setText(_translate("MainWindow", ""))
        self.label_tt.setText(_translate("MainWindow", ""))

        self.predict_button.setText(_translate("MainWindow", "Predict"))
        self.cancel_btn.setText(_translate("MainWindow", "Cancel"))
        self.c1.setItemText(0, _translate("MainWindow", "Yes"))
        self.c1.setItemText(1, _translate("MainWindow", "No"))
        self.c11.setItemText(0, _translate("MainWindow", "Yes"))
        self.c11.setItemText(1, _translate("MainWindow", "No"))
        self.c12.setItemText(0, _translate("MainWindow", "Yes"))
        self.c12.setItemText(1, _translate("MainWindow", "No"))
        self.c13.setItemText(0, _translate("MainWindow", "Yes"))
        self.c13.setItemText(1, _translate("MainWindow", "No"))
        self.c14.setItemText(0, _translate("MainWindow", "Yes"))
        self.c14.setItemText(1, _translate("MainWindow", "No"))
        self.c16.setItemText(0, _translate("MainWindow", "Yes"))
        self.c16.setItemText(1, _translate("MainWindow", "No"))
        self.c15.setItemText(0, _translate("MainWindow", "Male"))
        self.c15.setItemText(1, _translate("MainWindow", "Female"))
        self.label_35.setText(_translate("MainWindow", "Inflation rate: "))
        self.predict_button_2.setText(_translate("MainWindow", "Advanced Predict"))

    def predict(self):
        try:
            t1 = self.c1.currentText()
            t2 = self.d2.text()
            t3 = self.d3.text()
            t4 = self.d4.text()
            t5 = self.c5.currentText()
            t6 = self.d6.text()
            t7 = self.d7.text()
            t8 = self.d8.text()
            t9 = self.d9.text()
            t10 = self.d10.text()
            t11 = self.c11.currentText()
            t12 = self.c12.currentText()
            t13 = self.c13.currentText()
            t14 = self.c14.currentText()
            t15 = self.c15.currentText()
            t16 = self.c16.currentText()
            t17 = self.c17.text()
            t18 = self.c18.text()
            t19 = self.c19.text()
            t20 = self.c20.text()
            t21 = self.c21.text()
            t22 = self.c22.text()
            t23 = self.c23.text()
            t24 = self.c24.text()
            t25 = self.c25.text()
            t26 = self.c26.text()
            t27 = self.c27.text()
            t28 = self.c28.text()
            t29 = self.c29.text()
            t30 = self.c30.text()
            t31 = self.c31.text()
            t32 = self.c32.text()
            list_data = []
            list_data = [t1, t2, t3, t4, t5, t6, t7, t8, t9, t10, t11, t12, t13, t14, t15, t16, t17, t18, t19, t20, t21,
                         t22, t23, t24, t25, t26, t27, t28, t29, t30, t31, t32]

            for i in range(0, len(list_data)):
                if list_data[i] == '':
                    list_data[i] = 0
                if list_data[i] == 'Yes':
                    list_data[i] = 1
                elif list_data[i] == 'No':
                    list_data[i] = 0
                elif list_data[i] == 'Morning':
                    list_data[i] = 1
                elif list_data[i] == 'Evening':
                    list_data[i] = 0
                elif list_data[i] == 'Male':
                    list_data[i] = 1
                elif list_data[i] == 'Female':
                    list_data[i] = 0

            path = self.label_link.text()
            tt = self.label_tt.text()
            if tt == 'Random Forest Classifier':
                result = machine_learning.random_forest(path, list_data, int(self.max_depth),float(self.number))
            elif tt == 'Logistic Regression':
                float_list = list(map(float, list_data))
                result = machine_learning.logic_regression(path, float_list, float(self.number))
            elif tt == 'Linear Regression':
                result = machine_learning.liner(path, list_data, float(self.number))
            elif tt == 'Decision Tree Classifier':
                result = machine_learning.decision_tree(path, list_data, float(self.number), int(self.max_depth))
            else:
                # tt == 'K-Nearest Neighbors Classifier':
                float_list = list(map(float, list_data))
                result = machine_learning.knn(path, float_list, float(self.number), int(self.max_depth))
            if result[0] == 0:
                QMessageBox.information(self, 'Predict', 'Drop out', QMessageBox.Close)
            else:
                QMessageBox.information(self, 'Predict', 'Graduate', QMessageBox.Close)
        except Exception as err:
            QMessageBox.information(self, 'System',
                                    'An error occurred while predicting value. Please try again!\n'
                                    f'{str(err)}',
                                    QMessageBox.Close)

    def ad_predict(self):
        try:

            self.ui = QMainWindow()
            self.ui = advanced_prediction.UI_advanced_prediction(self.max_depth,self.number)
            self.ui.setupUi()
            self.ui.label_tt.setText(self.label_tt.text())
            self.ui.show()
            self.hide()
        except:
            QMessageBox.information(self, 'System',
                                    'An error occurred while opening advanced predict. Please try again!',
                                    QMessageBox.Close)

    def cancel(self):
        self.hide()


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    ui = Ui_predict_window()
    ui.setupUi()
    ui.show()
    sys.exit(app.exec_())
